airflow/models/dag.py

```diff
@@ -1269,8 +1269,8 @@ def get_task_instances_before(
         ``base_date``, or more if there are manual task runs between the
         requested period, which does not count toward ``num``.
         """
-        min_date = (
-            session.query(DagRun)
+        min_date: Optional[datetime] = (
+            session.query(DagRun.execution_date)
```

Comment:
This line is the only relevant fix; the rest are type annotations I added for debugging, they should be useful in the future as well.